src/Core/Content/Category/DataAbstractionLayer/CategoryIndexer.php

```diff
@@ -115,21 +116,31 @@ public function update(EntityWrittenContainerEvent $event): ?EntityIndexingMessa
             );
         }
 
-        $children = $this->fetchChildren($ids, $event->getContext()->getVersionId());
-        $ids = array_unique(array_merge($ids, $children));
+        $onlyParentIds = array_diff($parentIds, $ids);
+        foreach (array_chunk($onlyParentIds, self::UPDATE_IDS_CHUNK_SIZE) as $chunk) {
+            $parentIndexingMessage = new CategoryIndexingMessage($chunk, null, $event->getContext());
+            $parentIndexingMessage->setIndexer($this->getName());
+            EntityIndexerRegistry::addSkips($parentIndexingMessage, $event->getContext());
+            // for parents we only need to execute child count update
+            $parentIndexingMessage->addSkip(self::TREE_UPDATER, self::BREADCRUMB_UPDATER);
+
+            $this->messageBus->dispatch($parentIndexingMessage);
+        }
 
-        $chunks = \array_chunk($ids, self::UPDATE_IDS_CHUNK_SIZE);
-        $idsForReturnedMessage = array_shift($chunks);
+        $children = $this->fetchChildren($ids, $event->getContext()->getVersionId());
+        $onlyChildren = array_diff($children, $ids);
 
-        foreach ($chunks as $chunk) {
+        foreach (array_chunk($onlyChildren, self::UPDATE_IDS_CHUNK_SIZE) as $chunk) {
             $childrenIndexingMessage = new CategoryIndexingMessage($chunk, null, $event->getContext());
             $childrenIndexingMessage->setIndexer($this->getName());
             EntityIndexerRegistry::addSkips($childrenIndexingMessage, $event->getContext());
+            // for children we don't need to update the child count
+            $childrenIndexingMessage->addSkip(self::CHILD_COUNT_UPDATER);
 
             $this->messageBus->dispatch($childrenIndexingMessage);
         }
 
-        return new CategoryIndexingMessage($idsForReturnedMessage, null, $event->getContext());
+        return new CategoryIndexingMessage($ids, null, $event->getContext());
```

Comment:
we should still check if `$ids` needs to be chunked. as it could also be bigger than the chunk size, right?